lib/src/crawl_async.dart

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+final _empty = new Future.value(null);
+
+/// Finds and returns every node in a graph who's nodes and edges are
+/// asynchronously resolved.
+///
+/// Cycles are allowed. If this is an undirected graph the [children] function
+/// may be symmetric. In this case the [roots] may be any node in each connected
+/// graph.
+///
+/// [V] is the type of values in the graph nodes. [K] must be a type suitable
+/// for using as a Map or Set key. [children] should return the next reachable
+/// nodes.
+///
+/// There are no ordering guarantees. This is useful for ensuring some work is
+/// performed at every node in an asynchronous graph, but does not give
+/// guarantees that the work is done in topological order.
+Stream<V> crawlAsync<K, V>(Iterable<K> roots, FutureOr<V> Function(K) readNode,
+    FutureOr<Iterable<K>> Function(K, V) children) {
+  final crawl = new _CrawlAsync(roots, readNode, children);
+  crawl.run();
+  return crawl.result.stream;
+}
+
+class _CrawlAsync<K, V> {
+  final result = new StreamController<V>();
+
+  final FutureOr<V> Function(K) readNode;
+  final FutureOr<Iterable<K>> Function(K, V) children;
+  final Iterable<K> roots;
+
+  final _seen = new Set<K>();
+
+  _CrawlAsync(this.roots, this.readNode, this.children);
+
+  /// Add all nodes in the graph to [result] and return a Future which fires
+  /// after all nodes have been seen.
+  Future run() async {
+    await Future.wait(roots.map(_visit));
+    await result.close();
+  }
+
+  /// Resolve the node at [key] and output it, then start crawling all of it's
+  /// children.
+  Future _crawlFrom(K key) async {
+    var value = await readNode(key);
+    result.add(value);
+    var next = await children(key, value) ?? const [];
```

Comment:
Is supporting null here important? If we support it now then people will likely rely on that.

Reply:
I think it's a nice usability improvement. I have a test that we allow children to return null, and stronglyConnectedComponents behaves the same.